test/export/test_export.py

```diff
@@ -1680,6 +1674,67 @@ def forward(self, x):
         ):
             torch.export.export(exported_v2.module(), (torch.randn(2, 2),))
 
+    def test_cond_buffers(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_parameter("param", torch.nn.Parameter(torch.ones(2, 3), requires_grad=False))
+                self.register_buffer("buffer", torch.ones(2, 3) + 1)
+
+            def true_fn(self, x):
+                return x + self.param
+
+            def false_fn(self, x):
+                return x + self.buffer
+
+            def forward(self, x):
+                return cond(x.shape[0] == 4, self.true_fn, self.false_fn, [x])
+
+        inp = torch.ones(2, 3)
+        ep = torch.export.export(M(), (inp,))
+        inp = torch.randn(2, 3)
+        epm = ep.module()
+        self.assertTrue(torch.allclose(epm(inp), M()(inp)))
+
+        for gm in epm.named_modules():
+            if not isinstance(gm, torch.fx.GraphModule):
+                continue
+            self.assertEquals(
+                len([node for node in gm.graph.nodes if node.op == "placeholder"]),
+                1
+            )
+
+    def test_map_buffers(self):
```

Comment:
curious; was there a motivation for including this specific code pattern?

One thing I realized while basing some work off your diff is that I think we should ban this kind of code. In particular, code that references modules/parameters/buffers that are not in the module hierarchy.

If you define this module in eager, `param` and `buffer` will not be in `M`'s state dict, and `m1` will not be in the `children()`. All of this is very weird and unlikely to lead to correct results if you try to use this module for anything. So my claim is we should just error here.

Reply:
Oops sorry I missed this. No specific motivation for this code pattern. I just wanted a test case that uses a param/buffer inside map 😅